cmd/osbuild-service-maintenance/aws.go

```diff
@@ -192,9 +215,69 @@ func checkValidParent(childId string, parent []ec2types.Reservation) bool {
 	}
 
 	parentState := parent[0].Instances[0].State.Name
-	if parentState == ec2types.InstanceStateNameRunning || parentState == ec2types.InstanceStateNamePending {
+	if parentState != ec2types.InstanceStateNameTerminated {
 		return true
 	}
 	logrus.Infof("Instance %s has a parent (%s) in state %s, so we'll terminate %s.", childId, *parent[0].Instances[0].InstanceId, parentState, childId)
 	return false
 }
+
+func searchSGAndCleanup(ctx context.Context, a *awscloud.AWS, dryRun bool) error {
+	securityGroups, err := a.DescribeSecurityGroupsByPrefix(ctx, "SG for i-")
+	if err != nil {
+		return err
+	}
+
+	for _, sg := range securityGroups {
+		if sg.GroupId == nil || sg.GroupName == nil {
+			logrus.Errorf(
+				"Security Group needs to have a GroupId (%v) and a GroupName (%v).",
+				sg.GroupId,
+				sg.GroupName)
+			continue
+		}
+		reservations, err := a.DescribeInstancesBySecurityGroupID(*sg.GroupId)
+		if err != nil {
+			logrus.Errorf("Failed to describe security group %s: %v", *sg.GroupId, err)
+			continue
+		}
+
+		sgInUse := checkIfInUse(reservations)
+
+		// If no instance is running/pending, delete the SG
+		if !sgInUse {
+			logrus.Infof("Deleting security group: %s (%s)", *sg.GroupName, *sg.GroupId)
+			if !dryRun {
+				err := a.DeleteSecurityGroupById(ctx, sg.GroupId)
+
+				if err != nil {
+					logrus.Errorf("Failed to delete security group %s: %v", *sg.GroupId, err)
+				}
+			}
+		} else {
+			logrus.Debugf("Security group %s has non terminated instances associated with it.", *sg.GroupId)
+		}
+	}
+	return nil
+}
+
+// checkIfInUse returns true if any instance of the reservations is not terminated
+// then it's considered "in use"
+func checkIfInUse(reservations []ec2types.Reservation) bool {
```

Comment:
can we just call this "isInstanceRunning"? I thinkit makes it a bit clearer.

Reply:
It's more than running. Also pending, shutting-down.… but you are right, it sounds odd…

Reply:
or invert and `allTerminated`

Reply:
please check